Add HVV integration (Hamburg public transportation)

vigonotion opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I added a sensor for the HVV public transportation in Hamburg, Germany. It creates a sensor that shows the timestamp for the next departure, as well as the next (up to 20) departures in the attribute "next".
The integration is using the official api (Geofox API). You'll need credentials for the api, description on how to get them are in the docs pr.
It uses config flow.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: [https://github.com/home-assistant/home-assistant.io/pull/12002](https://github.com/home-assistant/home-assistant.io/pull/12002)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

vigonotion committed:
Fix linters

vigonotion committed:
tidy up consts

vigonotion committed:
fix api calls

vigonotion committed:
use pygti 0.4, full async

vigonotion committed:
Add log messages for network (un)reachable

vigonotion committed:
Use exceptions

vigonotion committed:
unlimited parallel updates

vigonotion committed:
Add options for config_flow

vigonotion committed:
Fix pylint warnings

vigonotion committed:
Add first test for config_flow

vigonotion committed:
Add invalid auth test on config flow

vigonotion committed:
Add cannotconnect test on config flow

vigonotion committed:
Add more tests

vigonotion committed:
Add attribution note

vigonotion committed:
Update documentation location

vigonotion committed:
Add no_results option

vigonotion committed:
Added files to coveragerc

vigonotion committed:
Fix pylint

springstan commented on cov.xml:
```diff
@@ -0,0 +1,260 @@
+<?xml version="1.0" ?>
```
Please remove this file.

springstan commented on homeassistant/components/hvv_departures/.translations/en.json:
```diff
@@ -0,0 +1,58 @@
+        },
+        "error": {
+            "cannot_connect": "Failed to connect, please try again",
+            "cn_too_many": "Too many results. Please precise your query",
```
`Precise` is not a verb 😁
```suggestion
            "cn_too_many": "Too many results. Please state your query more precisely",
```

springstan replied:
Please update this file as well.

springstan commented on homeassistant/components/hvv_departures/const.py:
```diff
@@ -0,0 +1,6 @@
+DOMAIN = "hvv_departures"
+DEFAULT_NAME = DOMAIN
+MANUFACTURER = "HVV"
+ATTRIBUTION = "Data provided by www.hvv.de."
```
```suggestion
ATTRIBUTION = "Data provided by www.hvv.de"
```

springstan commented on homeassistant/components/hvv_departures/manifest.json:
```diff
@@ -0,0 +1,16 @@
+  ],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
```
Please remove these lines since they are not used.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,215 @@
+ICON = "mdi:bus"
+UNIT_OF_MEASUREMENT = "min"
+
+ATTR_ATTRIBUTION = "attribution"
```
Import this constant from const.py.

vigonotion replied:
All of these are only used on the sensor platform. There is another binary_sensor platform planned, but they use a different icon and not the same attributes. See the developer documentation: https://developers.home-assistant.io/docs/en/next/creating_platform_code_review.html#0-common

As these constants are not widely used but just at platform level, they should be defined in sensor.py

springstan replied:
Sorry I was actually just referring to only `ATTR_ATTRIBUTION`.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,215 @@
+ATTR_ORIGIN = "origin"
+ATTR_DIRECTION = "direction"
+ATTR_TYPE = "type"
+ATTR_ID = "id"
```
Same as above.

springstan replied:
Only type and id.

vigonotion replied:
ATTR_TYPE is not in const.py

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,215 @@
+
+async def async_setup_platform(
+    hass, config, async_add_entities, discovery_info=None
+):  # pylint: disable=unused-argument
```
Why do we need to disable this?

vigonotion replied:
I think it was part of the generated code, removed it and pylint didn't complain

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,215 @@
+    Skipped, as setup through configuration.yaml is not supported.
+    """
+
+    pass
```
You can remove this line.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,215 @@
+    async def async_update(self, **kwargs):
+        """Update the sensor."""
+
+        try:
```
Only wrap the line where an error could occur.

springstan replied:
This has not been resolved.

vigonotion replied:
Understood it wrong, _should_ be correct now

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,223 @@
+    {
+        vol.Required("host", default=GTI_DEFAULT_HOST): str,
+        vol.Required("username"): str,
+        vol.Required("password"): str,
```
Please import these config keys from const.py.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,215 @@
+                self.config_entry.data["password"],
+                self.config_entry.data["host"],
+            )
+        )
```
What happens if the credentials are incorrect?

vigonotion replied:
That is checked inside async_update, where InvalidAuth is catched

vigonotion committed:
Fix precise is not a verb

vigonotion committed:
Remove dot from attribution

vigonotion committed:
Remove unused lines from manifest

vigonotion committed:
Improve async_setup_platform

vigonotion committed:
Fix Only wrap the line where an error could occur

vigonotion committed:
Use config keys from const.py

vigonotion committed:
Import attrs from const.py

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,214 @@
+                    self.config_entry.data["station"]["type"],
+                )
+            },
+            "name": self.config_entry.data["station"]["name"],
```
```suggestion
            "name": self._name,
```

vigonotion replied:
Actually, "Departures at Wartenau" is not the right device name. It should be the station name as before, as there will be a binary_sensor platform for the elevator status, and then the "Departures at.. " in the device name would be wrong

vigonotion replied:
I re-thought the whole naming thing. My integration is called "HVV Departures" and the main thing about this integration is the departures sensor. The upcoming elevator status is just supplementary information for the departure. So I'm gonna leave the device name like you suggested.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,214 @@
+        return ICON
+
+    @property
+    def available(self) -> bool:
```
```suggestion
    def available(self)
```

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,214 @@
+
+    def __init__(self, hass, config_entry, session):
+        """Initialize."""
+        self.hass = hass
```
`self.hass` is not used.

vigonotion committed:
Fix try catch

vigonotion committed:
Generate new translation file

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,214 @@
+        try:
+            data = await self.gti.departureList(payload)
+
+            if data["returnCode"] == "OK" and len(data.get("departures", [])) > 0:
```
You can simplify it by using the default return value of .get which is None and None is a falsy value so you can use it like this:
```suggestion
            if data["returnCode"] == "OK" and data.get("departures"):
```

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,213 @@
+                self._last_error = error
+            self._available = False
+
+        if data["returnCode"] == "OK" and len(data.get("departures", [])) > 0:
```
Use the default return value of .get() which is None and a falsy value like this:
```suggestion
        if data["returnCode"] == "OK" and data.get("departures"):
```

vigonotion committed:
Change device_name to reflect sensor name

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

vigonotion committed:
Remove type

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

vigonotion committed:
Use the default return value of .get()

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

vigonotion committed:
Fix available()

vigonotion committed:
Remove unused self.hass

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+                )
+
+            self.stations = {
+                "{} ({})".format(x.get("name"), x.get("type")): x for x in results
```
Please use f-strings see [style guidelines](https://developers.home-assistant.io/docs/en/next/development_guidelines.html#use-new-style-string-formatting).

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+                )
+
+                self.filters = {
+                    "{}, {}".format(x["serviceName"], x["label"]): x
```
Same as above.

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+        if "filter" in self.config_entry.options:
+            old_filter = "{}, {}".format(
+                self.config_entry.options["filter"][0]["serviceName"],
+                self.config_entry.options["filter"][0]["label"],
```
Same here.

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        errors = {}
+        if self.filters == {}:
```
Use it as a falsy value:
```suggestion
        if not self.filters:
```

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+        errors = {}
+        if self.filters == {}:
+
+            try:
```
Only wrap the line where an error could occur.

vigonotion replied:
Why is this a requirement? It makes it super complicated, as now I get "UnboundLocalError: local variable 'departure_list' referenced before assignment" errors and I have to check that before. Other integrations are also executing more than the falsy line inside the try.

```python
            try:
                departure_list = await self.hub.gti.departureList(
                    {
                        "station": self.config_entry.data["station"],
                        "time": {"date": "heute", "time": "jetzt"},
                        "maxList": 5,
                        "maxTimeOffset": 200,
                        "useRealtime": True,
                        "returnFilters": True,
                    }
                )
            except CannotConnect:
                errors["base"] = "cannot_connect"
            except InvalidAuth:
                errors["base"] = "invalid_auth"

            if departure_list:
                self.filters = {
                f"{x['serviceName']}, {x['label']}": x
                    for x in departure_list.get("filter")
                }
```

springstan replied:
Wrapping only the falsy line inside the try makes the code easier to read and maintain because you can directly see where an error could occur. I would recommend to declare the variable beforehand and then use it inside of the try.
Furthermore, I would refer back to the [documentation](https://developers.home-assistant.io/docs/en/next/creating_component_code_review.html) about other integrations not following this rule:

> Not all existing code follow the requirements in this checklist. This cannot be used as a reason to not follow them!

Moreover, multiple lines can be inside the try if the are falsy and return the same/similar exception.

vigonotion replied:
I also removed the authenticate call, as I noticed that everything is checked directly after that by departureList.

vigonotion replied:
I also removed the authenticate, because everything is checked by departureList directly after that

vigonotion committed:
Use f-strings instead of format

vigonotion committed:
Use not instead of {}

vigonotion committed:
Improve try in options flow

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
```
Could you please change it here as well?

vigonotion committed:
More try improvements

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+SCHEMA_STEP_OPTIONS = vol.Schema(
+    {
+        vol.Required("filter"): vol.In([]),
+        vol.Required("offset", default=0): vol.All(int, vol.Range(min=0)),
```
Please import and use `CONF_OFFSET` from const.py:
```suggestion
        vol.Required(CONF_OFFSET, default=0): vol.All(int, vol.Range(min=0)),
```

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+                    step_id="station", data_schema=SCHEMA_STEP_STATION, errors=errors
+                )
+
+            self.stations = {f"{x.get('name')} ({x.get('type')})": x for x in results}
```
Please rename `x` to something more self-explanatory, maybe station or something similar.

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+
+            self.filters = {
+                f"{x['serviceName']}, {x['label']}": x
+                for x in departure_list.get("filter")
```
Same as above.

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+                        self.filters.keys()
+                    ),
+                    vol.Required(
+                        "offset", default=self.config_entry.options.get("offset", 0)
```
```suggestion
                        CONF_OFFSET, default=self.config_entry.options.get("offset", 0)
```

vigonotion committed:
Use CONF_OFFSET from const.py

vigonotion committed:
Rename x to meaningful names

vigonotion committed:
Import config keys from const.py in tests

vigonotion committed:
Use cv.multi_select for filters

vigonotion commented:
As #31798 is now added, filters now use the new mutli_select option. The whole sensor logic was already prepared for that, I just had to tweak the schema in the options flow and correct the tests.

vigonotion commented:
PR is still open, waiting for approval

vigonotion committed:
Fix issue with missing timezone

vigonotion commented:
@springstan mind taking a look at it again?

springstan commented:
@vigonotion sure, sorry that I missed your review request.

I would suggest that you split this large PR into multiple ones for easier and faster reviews. It is quite common to only submit one platform for the first PR of a new integration. You can open other PRs for the other platforms to add support for them as well.

vigonotion commented:
I'm unsure where to split this PR, it only includes the sensor platform

springstan commented:
Oh yeah you are right 😅
Now I am unsure if it is possible to split up this large PR, @MartinHjelmare what do we usually do in such a case?

MartinHjelmare commented:
The code itself isn't so long, it's mostly the test fixture that eat lines. It's ok to keep it in one PR here.

springstan commented on homeassistant/components/hvv_departures/__init__.py:
```diff
@@ -0,0 +1,42 @@
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+PLATFORMS = ["sensor"]
```
Import the sensor domain from `sensor/init.py` and use it here.

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,241 @@
+    }
+)
+
+SCHEMA_STEP_STATION = vol.Schema({vol.Required("station"): str})
```
Please create a constant for `"station"` e.g. `CONF_STATION`

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,241 @@
+    {
+        vol.Required("filter"): vol.In([]),
+        vol.Required(CONF_OFFSET, default=0): vol.All(int, vol.Range(min=0)),
+        vol.Optional("realtime", default=True): bool,
```
Same as above and please split this word up: `"real_time"`

```suggestion
        vol.Optional(CONF_REAL_TIME, default=True): bool,
```

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,241 @@
+
+SCHEMA_STEP_OPTIONS = vol.Schema(
+    {
+        vol.Required("filter"): vol.In([]),
```
Same here. Btw why is the list empty for this config entry, is anything permitted here?

```suggestion
        vol.Required(CONF_FILTER): vol.In([]),
```

vigonotion replied:
Changed to CONF_FILTER.
It's an empty list because it should accept an array of json objects returned from the API. Should I specify the schema for that response in there?

```json
{
    "serviceID": "HHA-U:U1_HHA-U",
    "stationIDs": ["Master:10902"],
    "label": "Fuhlsbüttel Nord / Ochsenzoll / Norderstedt Mitte / Kellinghusenstraße / Ohlsdorf / Garstedt",
    "serviceName": "U1",
}
```

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,241 @@
+                    step_id="station", data_schema=SCHEMA_STEP_STATION, errors=errors
+                )
+
+        return self.async_show_form(
+            step_id="user", data_schema=SCHEMA_STEP_USER, errors=errors
+        )
```
I would recommend to move this to the top and only run this `if user_input is None:`

vigonotion replied:
Then the errors from line 62 won't have a return. I could maybe add another return at the top? But it shouldn't make a difference, just redundant code

springstan replied:
Just wanted to make it consistent with other config flows and IMHO it makes more sense to first display the form where the user can enter username and password and then go on after that.

vigonotion replied:
ok I put it on top

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,241 @@
+                step_id="station_select", data_schema=schema, errors=errors
+            )
+
+        return self.async_show_form(
+            step_id="station", data_schema=SCHEMA_STEP_STATION, errors=errors
+        )
```
Same as above, move this to the top and add a condtion before.

springstan commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,241 @@
+
+            return self.async_create_entry(title=title, data=self.data)
+
+        return self.async_show_form(
+            step_id="station_select", data_schema=SCHEMA_STEP_STATION, errors=errors
+        )
```
Same here, move it up and add a condition.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,213 @@
+                + timedelta(seconds=departure.get("delay", 0))
+            )
+
+            self.attr[ATTR_LINE] = departure["line"]["name"]
+            self.attr[ATTR_ORIGIN] = departure["line"]["origin"]
+            self.attr[ATTR_DIRECTION] = departure["line"]["direction"]
+            self.attr[ATTR_TYPE] = departure["line"]["type"]["shortInfo"]
+            self.attr[ATTR_ID] = departure["line"]["id"]
+            self.attr[ATTR_DELAY] = departure.get("delay", 0)
```
You can use `dict.update()` here:
```suggestion
            self.attr.update({
                ATTR_LINE: departure["line"]["name"],
                ATTR_ORIGIN: departure["line"]["origin"],
                ...
            })
```

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,213 @@
+            self._state = (
+                departure_time
+                + timedelta(minutes=departure["timeOffset"])
+                + timedelta(seconds=departure.get("delay", 0))
```
Create a local variable for the delay since you use it multiple time.

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,213 @@
+        station_id = self.config_entry.data["station"]["id"]
+        station_type = self.config_entry.data["station"]["type"]
+
+        return f"{DOMAIN}-{self.config_entry.entry_id}-{station_id}-{station_type}"
```
You can remove DOMAIN since unique_id only need to be unique for your integration.
```suggestion
        return f"{self.config_entry.entry_id}-{station_id}-{station_type}"
```

vigonotion replied:
I plan to add other platforms (binary_sensor) which share the same station_id and station_type, that's why I added the domain. Under these circumstances, can I keep the domain in the UUID?

springstan replied:
No, you can drop the domain in unique_ids because:

> An entity is looked up in the registry based on a combination of the plaform type (e.g., light), and the integration name (domain) (ie hue) and the unique ID of the entity.

source: https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

springstan commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,213 @@
+                        ATTR_DEPARTURE: departure_time
+                        + timedelta(minutes=departure["timeOffset"])
+                        + timedelta(seconds=departure.get("delay", 0)),
+                        ATTR_LINE: departure["line"]["name"],
```
Please create a local variable before appending the departure e.g. for `departure["line"]` since it is used so much here.

vigonotion replied:
just out of interest, is this change for ease of reading or are there performance boosts by using a local variable?

springstan replied:
Mostly for ease of reading, not sure how much the performance is improved by this change.

vigonotion committed:
Import the sensor domain from sensor/init.py and use it here.

vigonotion committed:
CONF_REAL_TIME

vigonotion committed:
CONF_FILTER

vigonotion committed:
Improve config flow station_select

vigonotion committed:
Use dict update

vigonotion committed:
local variable for departure['line']

vigonotion committed:
local variable for delay

vigonotion committed:
CONF_STATION 1/3

vigonotion committed:
CONF_STATION 2/3

vigonotion committed:
CONF_STATION 3/3

vigonotion committed:
Fix pylint

vigonotion committed:
remove DOMAIN from uuid

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

vigonotion committed:
user_input None on top

vigonotion committed:
user_input is None on top for step=station

vigonotion committed:
Update translations for realtime -> real_time change

MartinHjelmare commented:
Please rebase on latest dev branch and solve the merge conflict.

vigonotion committed:
Add HVV Departures (initial commit)

vigonotion committed:
Moved .translations to translations

vigonotion committed:
update translations

vigonotion committed:
Update manifest name

vigonotion commented:
Okay I rebased and also removed the dot from the translations folder

MartinHjelmare commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,249 @@
+
+    async def async_step_station(self, user_input=None):
+        """Handle the step where the user inputs his/her station."""
+        if user_input is None:
```
See above.

MartinHjelmare commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,249 @@
+        )
+
+        return self.async_show_form(
+            step_id="station_select", data_schema=schema, errors=errors
```
Please don't move to another step's form directly. The form and its errors are private to each step. Return and await the next step method instead.

MartinHjelmare commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,249 @@
+            departure_list = {}
+
+            session = aiohttp_client.async_get_clientsession(self.hass)
+            self.hub = GTIHub(
```
We should try to reuse the existing hub that the sensor is using. Save that in hass.data, then we can use it here in the options flow.

MartinHjelmare commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -0,0 +1,249 @@
+            options = {
+                CONF_FILTER: [
+                    departure_filter[1]
+                    for departure_filter in filter(
```
Please use a list comprehension instead of filter + lambda.

MartinHjelmare commented on homeassistant/components/hvv_departures/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "requirements": [
+    "pygti==0.6.0"
+  ],
+  "dependencies": [],
```
Please remove empty items.
```suggestion
```

MartinHjelmare commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,219 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+
+
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,219 @@
+ATTR_DELAY = "delay"
+ATTR_NEXT = "next"
+
+PARALLEL_UPDATES = 0
```
This means that there can be unlimited parallel updates.

vigonotion replied:
Yeah i guess that's what I want? I could just make up a number but I really don't know how to choose this number

MartinHjelmare replied:
If that's what we want, it's good. :+1:

MartinHjelmare commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,219 @@
+                self._last_error = error
+            self._available = False
+
+        if data["returnCode"] == "OK" and data.get("departures"):
```
Please invert this check and if true set available to false and return. Then we can outdent below.

MartinHjelmare commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,219 @@
+            delay = departure.get("delay", 0)
+            self._available = True
+            self._state = (
+                departure_time
```
The state for timestamp sensors should use absolute UTC iso-formatted time. Is that what this is? It doesn't seem like it's iso-formatted.

vigonotion replied:
I'm pretty sure it is I'm even using a HA method: from homeassistant.util.dt import utcnow
utcnow()

2020-06-08T16:17:04.397498+00:00 which is a valid iso time and HA can display it correctly

MartinHjelmare replied:
We want an isoformatted string representing the datetime.
```
Python 3.8.2 (default, Feb 26 2020, 02:56:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from homeassistant.util.dt import utcnow                                                                                                                                                              

In [2]: utcnow()                                                                                                                                                                                              
Out[2]: datetime.datetime(2020, 6, 8, 18, 2, 2, 823844, tzinfo=<UTC>)

In [3]: utcnow().isoformat()                                                                                                                                                                                  
Out[3]: '2020-06-08T18:02:05.869791+00:00'
```

MartinHjelmare commented on homeassistant/components/hvv_departures/strings.json:
```diff
@@ -0,0 +1,58 @@
+          "station": "Station/Address"
+        }
+      },
+      "options": {
```
There should be no options step in the main config flow.

MartinHjelmare commented on tests/components/hvv_departures/patched_data.py:
```diff
@@ -0,0 +1,194 @@
+"""Patched data for tests."""
```
Can we store these fixtures as json files instead under `/tests/fixtures/hvv_departures/`? We have a test helper to load fixtures in `tests/common.py`.

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,258 @@
+    flow.hass = hass
+
+    with patch(
+        "homeassistant.components.hvv_departures.config_flow.GTIHub.authenticate",
```
Please patch the library and not our code.
```suggestion
        "homeassistant.components.hvv_departures.hub.GTI.init",
```

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,258 @@
+"""Test the HVV Departures config flow."""
+from asynctest import patch
```
Please import patch from `tests.async_mock` instead.

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,258 @@
+
+async def test_user_flow(hass):
+    """Test that config flow works."""
+    flow = config_flow.ConfigFlow()
```
Please start the tests with the core interfaces. Use `hass.config_entries.flow.async_init`. I recommend using our scaffold script to generate example code with our current best practices.

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,258 @@
+    ), patch(
+        "pygti.gti.GTI.checkName",
+        return_value={
+            "returnCode": "OK",
```
Please store the api return values as json fixtures and load those into variables before use.

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,258 @@
+            user_input={CONF_STATION: "Wartenau (STATION)"}
+        )
+
+        print(result_station_select)
+
```
Stale print.
```suggestion
```

MartinHjelmare commented on tests/components/hvv_departures/test_hub.py:
```diff
@@ -0,0 +1,29 @@
+"""Test the HVV Departures GTIHub."""
```
I don't think we need this test it will be tested in the config flow tests.

MartinHjelmare commented on tests/components/hvv_departures/test_options_flow.py:
```diff
@@ -0,0 +1,192 @@
+"""Test the HVV Departures config flow."""
```
Please include these tests in the `test_config_flow.py` module. We want one test module per module under test.

vigonotion committed:
Address review requested changes

vigonotion committed:
add .isoformat()

vigonotion committed:
patch the library and not our code

vigonotion committed:
Use list comprehensions

vigonotion committed:
Make proper use of GTIHub

vigonotion committed:
Update homeassistant/components/hvv_departures/config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

vigonotion committed:
Update homeassistant/components/hvv_departures/config_flow.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

vigonotion committed:
Some more changes

vigonotion committed:
Merge branch 'vi/hvv' of https://github.com/vigonotion/home-assistant into vi/hvv

vigonotion committed:
Don't call other steps forms directly

vigonotion committed:
simplify schema creation

vigonotion committed:
Fix a bug that ADDRESS and POI (Point of interest) don't work for departures
A departureList can only return departures for a station, so ADDRESS and POI
will be excluded from the results.

vigonotion commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -107,22 +105,21 @@ async def async_step_station(self, user_input=None):
         """Handle the step where the user inputs his/her station."""
+
+        schema = vol.Schema(
+            {vol.Required(CONF_STATION): vol.In(list(self.stations.keys()))}
```
Didn't know that, fixed.

vigonotion committed:
Fix test for last commit

MartinHjelmare commented on homeassistant/components/hvv_departures/config_flow.py:
```diff
@@ -114,9 +115,7 @@ async def async_step_station(self, user_input=None):
-        schema = vol.Schema(
-            {vol.Required(CONF_STATION): vol.In(list(self.stations.keys()))}
-        )
+        schema = vol.Schema({vol.Required(CONF_STATION): vol.In(list(self.stations))})
```
We can remove `list` too as seen in my test.

vigonotion replied:
I had to undo that change, because otherwise this happens:

![image](https://user-images.githubusercontent.com/1506970/84538900-bf505f00-acf2-11ea-9072-be451be7b750.png)

Only if I add the list() around it, it works as intended:

![image](https://user-images.githubusercontent.com/1506970/84539058-0ccccc00-acf3-11ea-9807-26b8d584735a.png)

MartinHjelmare replied:
Right, that's a feature of the frontend that you can use a dictionary and have it render the values in the list for the user form while using the keys as return values.

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,380 @@
+    assert await hass.config_entries.async_setup(config_entry.entry_id)
+
+    with patch(
+        "homeassistant.components.hvv_departures.async_setup_entry", return_value=True
```
This patch has already been called at this point by `hass.config_entries.async_setup`.

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,380 @@
+                "serviceTypes": ["bus", "u"],
+                "hasStationInformation": True,
+            },
+            "stationInformation": {
```
Why do we store the station  information in the config entry? We should only store data that we need in the integration.

vigonotion replied:
It's not yet needed, I programmed the integration with a binary_sensor platform (for elevator status at that station) but for the smaller PR just included the sensor. But as soon as I PR the binary_sensor that information is needed and then that information is already available

MartinHjelmare replied:
Ok, then please only save that specific information. We don't need the return code, the station outline or the last update.

vigonotion replied:
I thought about it and I can probably get around with not saving that data in the config flow data (it could theoretically change anyways...) so i removed it completely.

MartinHjelmare commented on homeassistant/components/hvv_departures/sensor.py:
```diff
@@ -0,0 +1,201 @@
+    def __init__(self, hass, config_entry, session, hub):
+        """Initialize."""
+        self.config_entry = config_entry
+        self.station_name = self.config_entry.data["station"]["name"]
```
Please always use the available constants like `CONF_STATION`.

vigonotion committed:
remove list(...)

vigonotion committed:
remove redundant patch

vigonotion committed:
Use CONF_STATION

vigonotion committed:
Remove stationInformation from config flow data

vigonotion committed:
Undo removal of list()

MartinHjelmare commented on tests/components/hvv_departures/test_config_flow.py:
```diff
@@ -0,0 +1,346 @@
+    )
+    config_entry.add_to_hass(hass)
+
+    assert await hass.config_entries.async_setup(config_entry.entry_id)
```
We need to patch `"homeassistant.components.hvv_departures.hub.GTI"` before setting up the integration. We don't want any I/O in the tests.

MartinHjelmare commented on homeassistant/components/hvv_departures/translations/en.json:
```diff
@@ -0,0 +1,50 @@
+        },
+        "error": {
+            "cannot_connect": "Failed to connect, please try again",
+            "cn_too_many": "Too many results. Please state your query more precisely",
```
The `cn_too_many` key isn't used.

MartinHjelmare commented on homeassistant/components/hvv_departures/translations/en.json:
```diff
@@ -0,0 +1,50 @@
+            "cn_too_many": "Too many results. Please state your query more precisely",
+            "invalid_auth": "Invalid authentication",
+            "no_results": "No results. Try with a different station/address",
+            "unknown": "Unexpected error"
```
The `unknown` key isn't used.

vigonotion committed:
Remove unused strings

vigonotion committed:
Move config_entries.async_setup() into patched with

MartinHjelmare reviewed (approved):
Looks good!
